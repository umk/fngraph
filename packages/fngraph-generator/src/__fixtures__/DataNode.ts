import DataNode from '../DataNode'

export function createDataNode(overrides?: Partial<DataNode>): DataNode {
  return {
    incoming: [],
    outgoing: [],
    getter: jest.fn(),
    priority: 0,
    invert: false,
    getProperties: () => [],
    ...overrides,
  }
}
