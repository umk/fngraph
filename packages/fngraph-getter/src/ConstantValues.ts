import { DataRecord } from '@fngraph/data'
import {
  CONTEXT_GROUP_INVERSION,
  Falsy,
  GeneratorValue,
  Getter,
  GetterFactory,
} from '@fngraph/generator'

import { createIteration } from './IterationFunction'

type ConstantValues<V extends DataRecord> = Array<V | Falsy> | V | Falsy

export function createConstant<V extends DataRecord>(
  source: ConstantValues<V>,
): GetterFactory<never, V> {
  return function (_incoming, outgoing): Getter {
    const iterationF = createIteration(outgoing)
    return async function* (
      _properties,
      records,
      context,
      getRecord,
      invert,
    ): AsyncGenerator<GeneratorValue> {
      const iteration = iterationF(context, invert)
      for await (const current of records) {
        context.advanceToValueGroup(current)
        const base = getRecord(current.contexts)
        if (base) {
          const valuesArr = Array.isArray(source) ? source : [source]
          if (!invert) {
            yield* iteration(current, base, valuesArr.filter(Boolean) as Array<V>)
          } else {
            if (!valuesArr.some(Boolean)) {
              yield* iteration(current, base, CONTEXT_GROUP_INVERSION as V)
            }
          }
        }
      }
    }
  }
}

export default ConstantValues
