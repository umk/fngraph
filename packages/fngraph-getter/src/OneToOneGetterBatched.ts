import { DataRecord } from '@fngraph/data'
import {
  ContextGroup,
  DEFAULT_MAX_BATCH,
  DEFAULT_MAX_BUFFER,
  Falsy,
  GeneratorValue,
  Getter,
  GetterFactory,
  PropertyRef,
} from '@fngraph/generator'

import { createInversion } from './InversionFunction'
import { createIteration } from './IterationFunction'
import RecordBatchGroup, { getRecordBatchGroup } from './RecordBatchGroup'

type OneToOneGetterBatched<P extends DataRecord, R extends DataRecord> = (
  records: Array<P>,
  properties: Array<PropertyRef>,
) => Promise<Array<R | Falsy>>

export function createOneToOneBatched<P extends DataRecord, R extends DataRecord>(
  source: OneToOneGetterBatched<P, R>,
  maxBatch: number = DEFAULT_MAX_BATCH,
  maxBuffer: number = DEFAULT_MAX_BUFFER,
): GetterFactory<P, R> {
  return function (incoming, outgoing): Getter {
    const iterationF = createIteration(outgoing)
    return async function* (
      properties,
      records,
      context,
      getRecord,
      invert,
    ): AsyncGenerator<GeneratorValue> {
      const inversion = createInversion(invert)
      const iteration = iterationF(context, invert)
      for await (let value of records) {
        context.advanceToValueGroup(value)
        if (context.groups.length > context.groupIndex) {
          const derived = context.groups[context.groupIndex] as ContextGroup<R> | undefined
          if (derived) {
            // The cached derived record is defined only if and only
            // if the context is also defined.
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            const base = getRecord(value.contexts)!
            yield* iteration(value, base, derived as R)
          }
          context.groupIndex++
        } else {
          const groups = new Map<number, RecordBatchGroup<R>>()
          // The group indexes buffer maintains correct order of yield
          const groupIndexes: Array<number> = []
          let batchSize = 0 // The number of items to provide to the source
          for (let done: boolean | undefined = false; ; ) {
            if (!done) {
              const base = getRecord(value.contexts)
              if (!groups.has(context.groupIndex)) {
                const group = getRecordBatchGroup<R>(context, value, base)
                groups.set(context.groupIndex, group)
                if (group.include) batchSize++
              }
              groupIndexes.push(context.groupIndex)
              context.groupIndex++
            }
            if (groups.size === maxBatch || groupIndexes.length === maxBuffer || done) {
              if (batchSize > 0) {
                const g = Array.from(groups.values()).filter((g) => g.include)
                const derived = await source(
                  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                  g.map((g) => incoming(g.base!)),
                  properties,
                )
                derived.map(inversion).forEach((r, i) => (g[i].derived = r || undefined))
              }
              for (const groupIndex of groupIndexes) {
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                const group = groups.get(groupIndex)!
                if (context.isCached) context.groups[groupIndex] = group.derived
                if (group.derived) {
                  // The derived record is defined only when the base
                  // context is defined.
                  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                  yield* iteration(group.value, group.base!, group.derived as R)
                }
              }
              break
            }
            ;({ value, done } = await records.next())
            if (!done) context.advanceToValueGroup(value)
          }
        }
      }
    }
  }
}

export default OneToOneGetterBatched
