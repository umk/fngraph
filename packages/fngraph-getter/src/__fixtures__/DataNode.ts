import { DataNode } from '@fngraph/generator'

export function createDataNode(overrides?: Partial<DataNode>): DataNode {
  return {
    incoming: [],
    outgoing: [],
    getter: jest.fn(),
    priority: 0,
    invert: false,
    getProperties: () => [],
    ...overrides,
  }
}
