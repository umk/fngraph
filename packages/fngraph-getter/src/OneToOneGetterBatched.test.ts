import { DataRecord, DeclarationID } from '@fngraph/data'
import {
  contextAsRecord,
  createDataRecordGenerator,
  DataNodeSequence,
  recordAsContext,
} from '@fngraph/generator'

import { createConstant } from './ConstantValues'
import { createOneToOneBatched } from './OneToOneGetterBatched'
import { createDataNode } from './__fixtures__'

describe('createOneToOneBatched', () => {
  const declarationA = 'declaration A' as DeclarationID
  const declarationB = 'declaration B' as DeclarationID
  const declarationC = 'declaration C' as DeclarationID
  const declarationD = 'declaration D' as DeclarationID
  it('iterates through getter records', async () => {
    const values = new Array(10).fill(0).map((_, i) => ({
      [declarationA]: i * 2 + 1,
      [declarationB]: i * 2 + 2,
    }))
    const calls: Array<number> = []
    const nodeA = createDataNode({
      outgoing: [declarationA, declarationB],
      getter: createConstant(values)(contextAsRecord, recordAsContext),
    })
    const nodeB = createDataNode({
      incoming: [declarationA, declarationB],
      outgoing: [declarationC],
      getter: createOneToOneBatched((records) => {
        calls.push(records.length)
        const result = records.map((record) => {
          const valueA = record[declarationA] as number
          const valueB = record[declarationB] as number
          return { [declarationC]: valueA + valueB }
        })
        return Promise.resolve(result)
      }, 3)(contextAsRecord, recordAsContext),
    })
    const sequence = DataNodeSequence.create([nodeA, nodeB])
    const generator = createDataRecordGenerator(sequence, [])
    const records: Array<DataRecord> = []
    for await (const record of generator()) records.push(record)
    expect(records).toMatchSnapshot()
    expect(calls).toMatchSnapshot()
  })
  it('permutates records with another node', async () => {
    /**
     * Must result in the following evaluation sequence:
     *
     *   (B) <- (C) <- (A)
     *
     *        - or -
     *
     *   (B) <- (A) <- (C)
     */
    const values = new Array(10).fill(0).map((_, i) => ({
      [declarationA]: i * 2 + 1,
      [declarationB]: i * 2 + 2,
    }))
    const calls: Array<number> = []
    const nodeA = createDataNode({
      outgoing: [declarationA, declarationB],
      getter: createConstant(values)(contextAsRecord, recordAsContext),
    })
    const nodeB = createDataNode({
      incoming: [declarationA, declarationB],
      outgoing: [declarationC],
      getter: createOneToOneBatched((records) => {
        calls.push(records.length)
        const result = records.map((record) => {
          const valueA = record[declarationA] as number
          const valueB = record[declarationB] as number
          return { [declarationC]: valueA + valueB }
        })
        return Promise.resolve(result)
      }, 3)(contextAsRecord, recordAsContext),
    })
    const nodeC = createDataNode({
      outgoing: [declarationD],
      getter: createConstant([{ [declarationD]: 10 }, { [declarationD]: 20 }])(
        contextAsRecord,
        recordAsContext,
      ),
    })
    const sequence = DataNodeSequence.create([nodeA, nodeB, nodeC])
    const generator = createDataRecordGenerator(sequence, [])
    const records: Array<DataRecord> = []
    for await (const record of generator()) records.push(record)
    expect(records).toMatchSnapshot()
    expect(calls).toMatchSnapshot()
  })
})
