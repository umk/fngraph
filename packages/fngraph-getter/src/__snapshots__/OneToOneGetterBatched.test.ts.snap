// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createOneToOneBatched iterates through getter records 1`] = `
[
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 3,
  },
  {
    "declaration A": 3,
    "declaration B": 4,
    "declaration C": 7,
  },
  {
    "declaration A": 5,
    "declaration B": 6,
    "declaration C": 11,
  },
  {
    "declaration A": 7,
    "declaration B": 8,
    "declaration C": 15,
  },
  {
    "declaration A": 9,
    "declaration B": 10,
    "declaration C": 19,
  },
  {
    "declaration A": 11,
    "declaration B": 12,
    "declaration C": 23,
  },
  {
    "declaration A": 13,
    "declaration B": 14,
    "declaration C": 27,
  },
  {
    "declaration A": 15,
    "declaration B": 16,
    "declaration C": 31,
  },
  {
    "declaration A": 17,
    "declaration B": 18,
    "declaration C": 35,
  },
  {
    "declaration A": 19,
    "declaration B": 20,
    "declaration C": 39,
  },
]
`;

exports[`createOneToOneBatched iterates through getter records 2`] = `
[
  3,
  3,
  3,
  1,
]
`;

exports[`createOneToOneBatched permutates records with another node 1`] = `
[
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 3,
    "declaration D": 10,
  },
  {
    "declaration A": 3,
    "declaration B": 4,
    "declaration C": 7,
    "declaration D": 10,
  },
  {
    "declaration A": 5,
    "declaration B": 6,
    "declaration C": 11,
    "declaration D": 10,
  },
  {
    "declaration A": 7,
    "declaration B": 8,
    "declaration C": 15,
    "declaration D": 10,
  },
  {
    "declaration A": 9,
    "declaration B": 10,
    "declaration C": 19,
    "declaration D": 10,
  },
  {
    "declaration A": 11,
    "declaration B": 12,
    "declaration C": 23,
    "declaration D": 10,
  },
  {
    "declaration A": 13,
    "declaration B": 14,
    "declaration C": 27,
    "declaration D": 10,
  },
  {
    "declaration A": 15,
    "declaration B": 16,
    "declaration C": 31,
    "declaration D": 10,
  },
  {
    "declaration A": 17,
    "declaration B": 18,
    "declaration C": 35,
    "declaration D": 10,
  },
  {
    "declaration A": 19,
    "declaration B": 20,
    "declaration C": 39,
    "declaration D": 10,
  },
  {
    "declaration A": 1,
    "declaration B": 2,
    "declaration C": 3,
    "declaration D": 20,
  },
  {
    "declaration A": 3,
    "declaration B": 4,
    "declaration C": 7,
    "declaration D": 20,
  },
  {
    "declaration A": 5,
    "declaration B": 6,
    "declaration C": 11,
    "declaration D": 20,
  },
  {
    "declaration A": 7,
    "declaration B": 8,
    "declaration C": 15,
    "declaration D": 20,
  },
  {
    "declaration A": 9,
    "declaration B": 10,
    "declaration C": 19,
    "declaration D": 20,
  },
  {
    "declaration A": 11,
    "declaration B": 12,
    "declaration C": 23,
    "declaration D": 20,
  },
  {
    "declaration A": 13,
    "declaration B": 14,
    "declaration C": 27,
    "declaration D": 20,
  },
  {
    "declaration A": 15,
    "declaration B": 16,
    "declaration C": 31,
    "declaration D": 20,
  },
  {
    "declaration A": 17,
    "declaration B": 18,
    "declaration C": 35,
    "declaration D": 20,
  },
  {
    "declaration A": 19,
    "declaration B": 20,
    "declaration C": 39,
    "declaration D": 20,
  },
]
`;

exports[`createOneToOneBatched permutates records with another node 2`] = `
[
  3,
  3,
  3,
  1,
]
`;
